api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto

```diff
@@ -101,4 +102,13 @@ message DnsCacheConfig {
   // ``envoy.restart_features.use_apple_api_for_dns_lookups`` runtime value is true during
   // server startup. Apple' API only uses UDP for DNS resolution.
   bool use_tcp_for_dns_lookups = 8;
+
+  // If DNS resolvers are specified,
+  // DNS cache will perform DNS resolution via those resolvers.
+  // Setting this value causes failure if the
+  // ``envoy.restart_features.use_apple_api_for_dns_lookups`` runtime value is true during
+  // server startup. Apple's API only allows overriding DNS resolvers via system settings.
+  // If this setting is not specified, the value defaults to the default
+  // resolver, which uses /etc/resolv.conf for configuration.
+  repeated config.core.v3.Address dns_resolvers = 9;
```

Comment:
How does this relate to https://github.com/envoyproxy/envoy/pull/16237? Should this be a `repeated DnsResolver` and `DnsResolver` has both address and `DnsLookupOptions`?

Reply:
Thanks for letting me know about this change.

Looks like we're going to have `resolvers`, `use_tcp_for_dns_lookups` and `no_defalt_search_domain` as the arguments for `DnsResolverImpl` after #16237 is merged.
Seems like a good time to encapsulate those fields into a single proto message like `DnsResolver`.

@suniltheta How do you feel about the above changes?

Reply:
Thanks for letting me know about this PR. Just recently committed all of the changes for the PR #16237. A new protobuf message DnsResolverOptions in introduced which only include bool fields use_tcp_for_dns_lookups & no_default_search_domain.

Notice the name change from DnsLookupOptions to DnsResolverOptions.

Having said that if we move resolvers inside DnsResolverOptions, then we might have to take care of deprecating dns_resolvers present in config/cluster/v3/cluster.proto.

I lean towards keeping the resolvers (repeated Address type) out of dedicated DnsResolverOptions. Only because if we combine we are looking at a slightly bloated PR and deprecation of a field. But I am also flexible in combining them, if doing so is the best option in the long run.

Reply:
Yeah, I would separate `DnsResolverOptions` and reference it from a `DnsResolver` message, which can deal with addressing and other concerns not related to resolution options.

Reply:
So is this the format of the protobuf message that is being proposed?
```
└── DnsResolver
    └── <repeated> Address
    └── DnsResolverOptions
        ├── <bool> use_tcp_for_dns_lookups
        └── <bool> no_default_search_domain
        └── /<any options added in future>/
```

After the merge of #16237 we do
`dispatcher().createDnsResolver(resolvers, dns_resolver_options);`

Instead we would be doing
`dispatcher().createDnsResolver(dns_resolvers);`
So when we call the function `createDnsResolver ` we would be only passing the object of type `DnsResolver`. Here `dns_resolvers` will contain both `resolvers` and `dns_resolver_options`.

Reply:
Or is this the format
**FORMAT B**
```
├── DnsResolver
│   ├── <repeated Address> dns_resolvers
│   └── /<other concerns>/
└── <DnsResolverOptions> dns_resolver_options
        ├── <bool> use_tcp_for_dns_lookups
        ├── <bool> no_default_search_domain
        └── /<any options added in future>/
```
Finally the function call would be 
`dispatcher().createDnsResolver(dns_resolver, dns_resolver_options);`

Reply:
Sounds good. How do we want to proceed from here? I think we can either,

1) Wait for #16237 to be merged which adds `DnsResolverOptions` message and I can update the PR to create a new message `DnsResolver` and reference `DnsResolverOptions` from there.

2) Merge this PR as it is and have another PR that adds `DnsResolver ` message.

Reply:
I think we may not be able to merge #16237 or this PR without resolving the merge conflicts in the one which gets merged afterwards.
I am ok with either options, it is either one of us who would be taking care of merge conflicts. Let me know your preference and I can stick with that.

I am inclined to making "a new message `DnsResolver`" in a separate PR (different from this current PR) so that it will easy for review and nothing else.

Reply:
I think `FORMAT A:` is a better layout as it directly maps to a list of arguments one would need to construct a `DnsResolver`.

Reply:
@htuch Let us know if you're okay with the above plan. Thanks!

Reply:
Do we have `repeated DnsResolver` at top-level? I think each `DnsResolver` should have an `Address` and also a `DnsResolverOptions` perhaps.

Reply:
So instead of
```
└── DnsResolver
    ├── <repeated Address> dns_resolvers
    └── <DnsResolverOptions> dns_resolver_options
        ├── <bool> use_tcp_for_dns_lookups
        ├── <bool> no_default_search_domain
        └── /<any options added in future>/
```
We can have
```
└── repeated DnsResolver
    ├── <Address> dns_resolver
    └── <DnsResolverOptions> dns_resolver_options
        ├── <bool> use_tcp_for_dns_lookups
        ├── <bool> no_default_search_domain
        └── /<any options added in future>/
```
?
SGTM

Reply:
When we initialize the [DnsResolver](https://github.com/envoyproxy/envoy/blob/main/source/common/network/dns_impl.cc#L30) we have the ability to specify the DNS resolver lookup options once and send a list of external DNS servers (ip addresses) the resolver can query. I believe in the core implementation of c-ares resolver we call on this function [ares_set_servers_ports_csv](https://c-ares.haxx.se/ares_set_servers_ports_csv.html) on an already initialized channel to supply a string containing list of IP address. 

I do not see the benefit of specifying distinct resolver options for each external DNS server address. Please correct me if my assumptions are wrong.

Also @htuch is your suggestion on `repeated DnsResolver` is related to having the ability to specify the multiple DNS resolver libraries to be used? Like envoy currently uses c-ares as the core Dns Resolver and tomorrow there can be extensions added for a new Dns resolver (Eg: udns, adns etc). In that case should we add a field which says what DNS resolver library to be initialized (c-ares or other DNS resolver extensions) ?

> Sorry about dragging the conversation on a PR comment. Should we create an github issue (design proposal) and discuss there?

Reply:
Yeah, fair point. Let's go with https://github.com/envoyproxy/envoy/pull/16294#discussion_r627609104 then.

Reply:
@htuch What's the next step? @suniltheta and I discussed merging this PR as it is and make another PR to add the new `DnsResolver` message.

Reply:
If we're planning on immediately restructuring in the next PR that isn't so great. Can we at least introduce the `DnsResolver` message and place the list of addresses in there?

Reply:
I am on board with this idea as well. We can place the `DnsResolver` protobuf message in `api/envoy/config/core/v3/base.proto` or any other place which makes more sense in this PR. Move `repeated Address dns_resolvers` inside `DnsResolver` like mentioned in previous comment. Finally as part of #16237 PR we can move the `DnsResolverOptions` inside the already merged `DnsResolver` message.

Reply:
+1